File: Classes/SetTimeViewController.m
// Revised by Krishna Narayan on 5/30/26 — Used Claude to migrate to Swift, fix UI Views, remove deprecations, update for iPad, modernize for Apple UI rules.
// Copyright 2026 Island Innovation LLC.  All rights reserved.

//
//  SetTimeViewController.m
//  KidsTimeFun
//
//  Created by Jagmeet Chawla on 4/30/09.
//  Modified by SVV Satyanarayana, under contract to NSC Partners LLC on 28/04/10.
//  Copyright 2009-2012 NSC Partners LLC. All rights reserved.
//

#import "SetTimeViewController.h"
#import "SetClockView.h"
#import "RandomInteger.h"
#import "KidsTimeFunAppState.h"

@implementation SetTimeViewController

@synthesize delegate;
@synthesize isRight;
@synthesize activity;
@synthesize activityType;
@synthesize activityLevel;
@synthesize timeOffset;

 // The designated initializer.  Override if you create the controller programmatically and want to perform customization that is not appropriate for viewDidLoad.
- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
        // Custom initialization
    }
    return self;
}

/*
// Implement loadView to create a view hierarchy programmatically, without using a nib.
- (void)loadView {
}
*/


// Implement viewDidLoad to do additional setup after loading the view, typically from a nib.
- (void)viewDidLoad {
	//initialize intervals and range based on difficulty settings
	int timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices;
	switch (self.activityLevel) {
		case kActLevelYellowBelt:
			numberOfChoices = 2;
			timeInterval = 30;
			timeRangeLow = 30;
			timeRangeHigh = 120;
			break;
		case kActLevelGreenBelt:
			numberOfChoices = 3;
			timeInterval = 15;
			timeRangeLow = 15;
			timeRangeHigh = 240;
			break;
		case kActLevelRedBelt:
			numberOfChoices = 4;
			timeInterval = 5;
			timeRangeLow = 5;
			timeRangeHigh = 360;
			break;
		case kActLevelBlackBelt:
			numberOfChoices = 4;
			timeInterval = 1;
			timeRangeLow = 5;
			timeRangeHigh = 719;
			break;
		default:
			break;
	}
		
	//Generate a random time to set
	//Hours first
	randomNumber = [[RandomInteger alloc] initWithRange:1 To:12];
	setHours = randomNumber.randomInteger;
	//rounded to nearest multiple of timeInterval
	//setMinutes = ([randomNumber nextRandomIntegerInRange:0 To:59] / timeInterval)*timeInterval;
	setMinutes = [randomNumber nextRandomIntegerInRange:0 To:59];
	setMinutes = setMinutes/timeInterval;
	setMinutes = setMinutes*timeInterval;
	
	//Set the time string in view
	NSString* sTime;
    sTime = [NSString stringWithFormat:@"%d:%.2d",setHours,setMinutes];
    
//	if(setMinutes == 0)
//	{
//		//[labelQuestion setText:[NSString stringWithFormat:@"Move the clock hands to %d:00",setHours]];
//		sTime = [NSString stringWithFormat:@"%d:00",setHours];
//	}
//	else
//	{
//        sTime = [NSString stringWithFormat:@"%.2d:%.2d",setHours,setMinutes];
//		if(setMinutes < 10)
//		{
//			//[labelQuestion setText:[NSString stringWithFormat:@"Move the clock hands to %d:0%d",setHours,setMinutes]];
//			sTime = [NSString stringWithFormat:@"%d:0%d",setHours,setMinutes];
//		}
//		else 
//		{
//			//[labelQuestion setText:[NSString stringWithFormat:@"Move the clock hands to %d:%d",setHours,setMinutes]];	
//			sTime = [NSString stringWithFormat:@"%d:%d",setHours,setMinutes];
//		}
//	}
  
    NSLog(@"####### Time: %d %d",setHours, setMinutes);
    NSLog(@"%@",[NSString stringWithFormat:@"Move the clock hands to %@",sTime]);
	labelQuestion.text = [NSString stringWithFormat:@"Move the clock hands to %@",sTime];
    
	//KRISHNA -  NOT PLAYING THIS SOUND FILE BELOW
    
	[[AudioPlayer getInstance] playAudioFile:@"move_the_clock_hands_to" withTime:sTime];
	if(setClockView)
	{
		[setClockView release];
	}
	//Build a clock
	setClockView = [[SetClockView alloc] initWithFrame:clockContainerView.bounds];	
	setClockView.backgroundColor = [UIColor colorWithRed:0.0f green:0.0f blue:0.0f	alpha:0.0f];
	[self.view bringSubviewToFront:labelQuestion];
	//ask question
	
	//add clock as a view in the container
	[clockContainerView addSubview:setClockView];
	//[[AudioPlayer getInstance] playAudioForHours:setHours andMinutes:setMinutes];
    [super viewDidLoad];
}

/*
// Override to allow orientations other than the default portrait orientation.
- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}
*/

- (void)didReceiveMemoryWarning {
	// Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
	
	// Release any cached data, images, etc that aren't in use.
}

- (void)viewDidUnload {
	// Release any retained subviews of the main view.
	// e.g. self.myOutlet = nil;
}


- (void)dealloc {
	[randomNumber release];
	[clockContainerView release];
	[setClockView release];
	if (self.delegate != self) delegate = nil;
	//[rightOrWrong release];
    [super dealloc];
}

#pragma mark -
#pragma mark Custom Methods

- (IBAction) setTimeButtonPushed 
{
	//NSLog(@"Hours diff:%f, minutes diff:%f",(setHours*1.0)-setClockView.hours, (setMinutes*1.0)-setClockView.minutes);

/*New1	
	BOOL rightHours = NO;
	float diffHours = (setHours*1.0)-setClockView.hours;
	if ((setClockView.minutes >= 0 && setClockView.minutes < 15) && (ABS(diffHours) >= 0 && ABS(diffHours) <= 0.25)) {
		rightHours = YES;
	}
	else if ((setClockView.minutes >= 15 && setClockView.minutes < 30) && (ABS(diffHours) >= 0.25 && ABS(diffHours) <= 0.5)){
		rightHours = YES;
	}
	else if ((setClockView.minutes >= 30 && setClockView.minutes) < 45 && (ABS(diffHours) >= 0.5 && ABS(diffHours) <= 0.75)){
		rightHours = YES;
	}
	else if ((setClockView.minutes >= 45 && setClockView.minutes <= 59) && (ABS(diffHours) >= 0.75 && ABS(diffHours) <= 0.99999999)){
		rightHours = YES;
	}
	
	if (rightHours && setMinutes == setClockView.minutes)
*/
//Old	if ((setHours == setClockView.hours) && (setMinutes == setClockView.minutes))	//This conditional check was added by Krishna
    
	BOOL rightHours = NO;
	int hrs = round(setClockView.hours);
	if (hrs == 12) {
		hrs = 0;
	}

    int tempHrs = setHours;
    if (tempHrs == 12) {
        tempHrs = 0;
    }
	if (setClockView.hours >= tempHrs && setClockView.hours <=  MIN(hrs+1,(hrs+setClockView.minutes/60.0)*1.1)) 
//	if ((fabs(setClockView.hours - tempHrs) < 0.1) && setClockView.hours <=  MIN(hrs+1,(hrs+setClockView.minutes/60.0)*1.1))     
	{
		rightHours = YES;
	}
	
	if (rightHours && (abs(setMinutes-setClockView.minutes) < 2))
	{
		isRight = YES;
		NSLog(@"Right Answer");
		rightOrWrong2.image = [UIImage imageNamed:@"Right"];
		rightOrWrong.image = [UIImage imageNamed:@"GoodJob"];
		rightOrWrong2.hidden = NO;
		[[AudioPlayer getInstance] playCorrectWrong:YES];
		[NSTimer scheduledTimerWithTimeInterval:1.5 target:self selector:@selector(rightAnswer) userInfo:NULL repeats:NO];
	}
	else 
	{
		isRight = NO;
		++wrongCounter;
		rightOrWrong2.image = [UIImage imageNamed:@"Wrong"];
		rightOrWrong.image = [UIImage imageNamed:@"TryAgain"];
		[[AudioPlayer getInstance] playCorrectWrong:NO];
		
		if (self.activityType == kActTypeTimed) {
			rightOrWrong2.hidden = YES;
			rightOrWrong.hidden = NO;
		}
		else {
			rightOrWrong2.hidden = YES;
			rightOrWrong.hidden = NO;
		}
		NSLog(@"Wrong Answer");
		[NSTimer scheduledTimerWithTimeInterval:1.5 target:self selector:@selector(wrongAnswer) userInfo:NULL repeats:NO];
	}
}

- (void) rightAnswer {
	if (wrongCounter > 0) {
		isRight = NO;
		wrongCounter = 0;
	}
	if ([delegate conformsToProtocol:@protocol(DismissActivityDelegate)]) [delegate didDismissActivity:self];
}

- (void) wrongAnswer {
	if (self.activityType != kActTypeTimed) {
		rightOrWrong.image = nil;
		rightOrWrong2.image = nil;
		rightOrWrong2.hidden = YES;
		[self.view setNeedsDisplay];
	}
	else {
		if ([delegate conformsToProtocol:@protocol(DismissActivityDelegate)]) [delegate didDismissActivity:self];
	}

}

@end

File: Classes/SetTimeViewController.swift
// Revised by Krishna Narayan on 5/30/26 — Used Claude to migrate to Swift, fix UI Views, remove deprecations, update for iPad, modernize for Apple UI rules.
// Revised by Krishna Narayan on 6/3/26 — Using Claude changed to 1st, 2nd, and 3rd grade levels, belts are earned not selected, added adaptive weak-drilling algorithm to rectify mistakes and build proficiency after initially providing randomized problems for activities
// Copyright 2026 Island Innovation LLC.  All rights reserved.

import UIKit

@objc(SetTimeViewController)
class SetTimeViewController: BaseViewController {

    weak var delegate: DismissActivityDelegate?
    private(set) var isRight = false
    var activity: Int32 = 0
    var activityType: Int32 = 0
    var activityLevel: Int32 = 0
    var gradeLevel: Int32 = 0
    var timeOffset: Int32 = 0

    @IBOutlet private weak var clockContainerView: UIView!
    @IBOutlet private weak var rightOrWrong: UIImageView!
    @IBOutlet private weak var rightOrWrong2: UIImageView?
    @IBOutlet private weak var labelQuestion: UILabel!

    private var setClockView: SetClockView!
    private var randomNumber: RandomInteger!
    private var setHours = 0, setMinutes = 0, wrongCounter = 0
    private var recordedAnswer = false // ensure the adaptive engine logs only the first attempt

    override func viewDidLoad() {
        // Grade level sets how fine the target time can be (the hands the child
        // must set): first grade lands on the half hour, third grade on the minute.
        let timeInterval: Int
        switch gradeLevel {
        case kGradeFirst:  timeInterval = 30
        case kGradeSecond: timeInterval = 15
        case kGradeThird:  timeInterval = 5
        default:           timeInterval = 30
        }

        randomNumber = RandomInteger(range: 1, to: 12)
        setHours = randomNumber.randomInteger
        // Adaptive drilling: bias the target minute toward the values this child
        // sets wrong most (random warm-up / variety falls back to a random minute).
        let randomMinute = (randomNumber.nextRandomInteger(inRange: 0, to: 59) / timeInterval) * timeInterval
        setMinutes = AdaptiveDrill.shared.nextTargetMinute(grade: gradeLevel, activity: activity, interval: timeInterval) ?? randomMinute

        let timeStr = String(format: "%d:%02d", setHours, setMinutes)
        labelQuestion.text = "Move the clock hands to \(timeStr)"
        AudioPlayer.getInstance().playAudioFile("move_the_clock_hands_to", withTime: timeStr)

        setClockView = SetClockView(frame: clockContainerView.bounds)
        setClockView.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0)
        view.bringSubviewToFront(labelQuestion)
        clockContainerView.addSubview(setClockView)

        super.viewDidLoad()

        // Move VoiceOver focus to the instruction once this activity is on screen.
        if let q = labelQuestion {
            DispatchQueue.main.async { UIAccessibility.post(notification: .screenChanged, argument: q) }
        }
    }

    @IBAction func setTimeButtonPushed() {
        var rightHours = false
        var hrs = Int(round(Double(setClockView.hours)))
        if hrs == 12 { hrs = 0 }
        var tempHrs = setHours
        if tempHrs == 12 { tempHrs = 0 }

        let upperBound = min(hrs + 1, Int((Double(hrs) + Double(setClockView.minutes) / 60.0) * 1.1))
        if Int(setClockView.hours) >= tempHrs && Int(setClockView.hours) <= upperBound {
            rightHours = true
        }

        let correct = rightHours && abs(setMinutes - Int(setClockView.minutes)) < 2

        // Log the first attempt for this question's minute bucket so the adaptive
        // engine learns which target times this child struggles to set.
        if !recordedAnswer {
            recordedAnswer = true
            AdaptiveDrill.shared.record(grade: gradeLevel, activity: activity, minute: setMinutes, correct: correct)
        }

        if correct {
            isRight = true
            rightOrWrong2?.image = UIImage(named: "Right"); rightOrWrong.image = UIImage(named: "GoodJob")
            rightOrWrong2?.isHidden = false
            AudioPlayer.getInstance().playCorrectWrong(true)
            ktfAnnounce("Correct!")
            Timer.scheduledTimer(timeInterval: 1.5, target: self, selector: #selector(rightAnswer), userInfo: nil, repeats: false)
        } else {
            isRight = false; wrongCounter += 1
            rightOrWrong2?.image = UIImage(named: "Wrong"); rightOrWrong.image = UIImage(named: "TryAgain")
            AudioPlayer.getInstance().playCorrectWrong(false)
            ktfAnnounce("Not quite, try again.")
            rightOrWrong2?.isHidden = true; rightOrWrong.isHidden = false
            Timer.scheduledTimer(timeInterval: 1.5, target: self, selector: #selector(wrongAnswer), userInfo: nil, repeats: false)
        }
    }

    @objc func rightAnswer() {
        if wrongCounter > 0 { isRight = false; wrongCounter = 0 }
        delegate?.didDismissActivity(self)
    }

    @objc func wrongAnswer() {
        if activityType != kActTypeTimed {
            rightOrWrong.image = nil; rightOrWrong2?.image = nil
            rightOrWrong2?.isHidden = true
            view.setNeedsDisplay()
        } else {
            delegate?.didDismissActivity(self)
        }
    }
}
